File: src/orchestrator/workspace-engine.js
import { createDefaultConfig } from '../config/defaults.js';
import { AuditLog } from '../audit/audit-log.js';
import { SecretScanner } from '../security/secret-scanner.js';
import { PathPolicy } from '../workspace/path-policy.js';
import { FileAccessController } from '../workspace/file-access.js';
import { ProjectIndexer } from '../workspace/indexer.js';
import { ContextManager } from '../context/context-manager.js';
import { CommandPolicy } from '../commands/command-policy.js';
import { CommandRunner } from '../commands/command-runner.js';
import { GitService } from '../git/git-service.js';
import { PatchEngine } from '../patch/patch-engine.js';
import { OpenAICompatibleAdapter } from '../model/openai-compatible-adapter.js';

export class WorkspaceEngine {
  constructor({ config, scanner, auditLog, pathPolicy, fileAccess, indexer, contextManager, commandPolicy, commandRunner, git, patchEngine, modelAdapter }) {
    this.config = config;
    this.scanner = scanner;
    this.auditLog = auditLog;
    this.pathPolicy = pathPolicy;
    this.fileAccess = fileAccess;
    this.indexer = indexer;
    this.contextManager = contextManager;
    this.commandPolicy = commandPolicy;
    this.commandRunner = commandRunner;
    this.git = git;
    this.patchEngine = patchEngine;
    this.modelAdapter = modelAdapter;
  }
}

export function createDefaultEngine(overrides = {}) {
  const config = createDefaultConfig(overrides.config ?? {});
  const scanner = overrides.scanner ?? new SecretScanner({ customPatterns: config.secretPatterns });
  const auditLog = overrides.auditLog ?? new AuditLog({
    dataDir: config.dataDir,
    enabled: config.audit.enabled,
    scanner
  });
  const pathPolicy = overrides.pathPolicy ?? new PathPolicy({
    allowedRoots: config.allowedRoots,
    restrictedPatterns: config.restrictedPatterns
  });
  const fileAccess = overrides.fileAccess ?? new FileAccessController({ config, auditLog, scanner, pathPolicy });
  const indexer = overrides.indexer ?? new ProjectIndexer({ config, scanner, auditLog });
  const contextManager = overrides.contextManager ?? new ContextManager({ fileAccess, scanner });
  const commandPolicy = overrides.commandPolicy ?? new CommandPolicy({ config });
  const commandRunner = overrides.commandRunner ?? new CommandRunner({ config, commandPolicy, auditLog, scanner });
  const git = overrides.git ?? new GitService({ auditLog });
  const patchEngine = overrides.patchEngine ?? new PatchEngine({ config, auditLog, scanner, pathPolicy });
  const modelAdapter = overrides.modelAdapter ?? new OpenAICompatibleAdapter({
    baseUrl: config.model.baseUrl,
    apiKey: process.env[config.model.apiKeyEnv],
    model: config.model.model,
    timeoutMs: config.model.timeoutMs
  });

  return new WorkspaceEngine({
    config,
    scanner,
    auditLog,
    pathPolicy,
    fileAccess,
    indexer,
    contextManager,
    commandPolicy,
    commandRunner,
    git,
    patchEngine,
    modelAdapter
  });
}

File: crates/workspace-engine/src/workspace_engine.rs
use crate::audit::AuditLog;
use crate::chat::ChatOrchestrator;
use crate::command_policy::CommandPolicy;
use crate::command_runner::CommandRunner;
use crate::config::Config;
use crate::context_manager::ContextManager;
use crate::file_access::FileAccessController;
use crate::git_service::GitService;
use crate::indexer::ProjectIndexer;
use crate::patch_engine::PatchEngine;
use crate::path_policy::PathPolicy;
use crate::secret_scanner::SecretScanner;
use crate::session::SessionStore;

#[derive(Debug, Clone)]
pub struct WorkspaceEngine {
    pub config: Config,
    pub scanner: SecretScanner,
    pub audit_log: AuditLog,
    pub path_policy: PathPolicy,
    pub file_access: FileAccessController,
    pub indexer: ProjectIndexer,
    pub context_manager: ContextManager,
    pub command_policy: CommandPolicy,
    pub command_runner: CommandRunner,
    pub git: GitService,
    pub patch_engine: PatchEngine,
    pub session_store: SessionStore,
    pub chat_orchestrator: ChatOrchestrator,
}

impl WorkspaceEngine {
    pub fn new(config: Config) -> Self {
        let scanner = SecretScanner::default();
        let audit_log = AuditLog::new(&config.data_dir, true, scanner.clone());
        let path_policy = PathPolicy::new(&config);
        let file_access = FileAccessController::new(
            config.clone(),
            audit_log.clone(),
            scanner.clone(),
            path_policy.clone(),
        );
        let indexer = ProjectIndexer::new(config.clone(), scanner.clone(), audit_log.clone());
        let context_manager = ContextManager::new(file_access.clone(), scanner.clone());
        let command_policy = CommandPolicy::new(config.clone());
        let command_runner = CommandRunner::new(
            config.clone(),
            command_policy.clone(),
            audit_log.clone(),
            scanner.clone(),
        );
        let git = GitService::new(audit_log.clone());
        let patch_engine = PatchEngine::new(
            config.clone(),
            audit_log.clone(),
            scanner.clone(),
            path_policy.clone(),
        );
        let session_store = SessionStore::new(&config.data_dir);
        let chat_orchestrator = ChatOrchestrator::new(
            config.clone(),
            scanner.clone(),
            audit_log.clone(),
            indexer.clone(),
            context_manager.clone(),
            session_store.clone(),
        );

        Self {
            config,
            scanner,
            audit_log,
            path_policy,
            file_access,
            indexer,
            context_manager,
            command_policy,
            command_runner,
            git,
            patch_engine,
            session_store,
            chat_orchestrator,
        }
    }
}

impl Default for WorkspaceEngine {
    fn default() -> Self {
        Self::new(Config::default())
    }
}
